applications/CompressiblePotentialFlowApplication/custom_elements/transonic_perturbation_potential_flow_element.cpp

```diff
@@ -0,0 +1,856 @@
+        rRightHandSideVector[rRow] = rUpper_rhs(rRow);
+        rRightHandSideVector[rRow + NumNodes] = rWake_rhs(rRow);
+    }
+    else{
+        rRightHandSideVector[rRow] = rWake_rhs(rRow);
+        rRightHandSideVector[rRow + NumNodes] = rLower_rhs(rRow);
+    }
+}
+
+template <int Dim, int NumNodes>
+void TransonicPerturbationPotentialFlowElement<Dim, NumNodes>::ComputePotentialJump(const ProcessInfo& rCurrentProcessInfo)
+{
+    const array_1d<double, 3>& vinfinity = rCurrentProcessInfo[FREE_STREAM_VELOCITY];
+    const double vinfinity_norm = sqrt(inner_prod(vinfinity, vinfinity));
+
+    array_1d<double, NumNodes> distances;
+    GetWakeDistances(distances);
+
+    for (unsigned int i = 0; i < NumNodes; i++)
+    {
+        double aux_potential =
+            GetGeometry()[i].FastGetSolutionStepValue(AUXILIARY_VELOCITY_POTENTIAL);
+        double potential = GetGeometry()[i].FastGetSolutionStepValue(VELOCITY_POTENTIAL);
+        double potential_jump = aux_potential - potential;
+
+        if (distances[i] > 0)
+        {
+            GetGeometry()[i].SetValue(POTENTIAL_JUMP,
+                                      -2.0 / vinfinity_norm * (potential_jump));
+        }
+        else
+        {
+            GetGeometry()[i].SetValue(POTENTIAL_JUMP, 2.0 / vinfinity_norm * (potential_jump));
+        }
+    }
+}
+
+template <int Dim, int NumNodes>
+double TransonicPerturbationPotentialFlowElement<Dim, NumNodes>::ComputeDensity(const ProcessInfo& rCurrentProcessInfo) const
+{
+    // Reading free stream conditions
+    const double rho_inf = rCurrentProcessInfo[FREE_STREAM_DENSITY];
+    const double M_inf = rCurrentProcessInfo[FREE_STREAM_MACH];
+    const double heat_capacity_ratio = rCurrentProcessInfo[HEAT_CAPACITY_RATIO];
+    const double mach_number_limit = rCurrentProcessInfo[MACH_LIMIT];
+
+    // Computing local mach number
+    double local_mach_number = PotentialFlowUtilities::ComputePerturbationLocalMachNumber<Dim, NumNodes>(*this, rCurrentProcessInfo);
+
+    if (local_mach_number > mach_number_limit)
+    { // Clamping the mach number to mach_number_limit
+        KRATOS_WARNING("ComputeDensity") << "Clamping the local mach number to " << mach_number_limit << std::endl;
+        local_mach_number = mach_number_limit;
+    }
+
+    // Computing squares
+    const double M_inf_2 = M_inf * M_inf;
+    const double M_2 = local_mach_number * local_mach_number;
+
+    // Computing density according to Equation 8.9 of Drela, M. (2014) Flight Vehicle
+    // Aerodynamics, The MIT Press, London
+    const double numerator = 1 + (heat_capacity_ratio - 1) * M_inf_2 / 2;
+    const double denominator = 1 + (heat_capacity_ratio - 1) * M_2 / 2;
+    const double base = numerator / denominator;
+
+    if (base > 0.0)
+    {
+        return rho_inf * pow(base, 1 / (heat_capacity_ratio - 1));
+    }
+    else
+    {
+        KRATOS_WARNING("ComputeDensity") << "Using density correction" << std::endl;
+        return rho_inf * 0.00001;
```

Comment:
I'm wondering if this tolerance should be exposed to the user through the `ProcessInfo`.

Reply:
Actually you are raising a good point here. This implementation follows the method presented in _D. Eller. “Fast, Unstructured-Mesh Finite-Element Method for Nonlinear Subsonic Flow”. In: Journal of aircraft 49.5 (2012), pp. 1471–1479._ where it is explained and I simply quote: "Because of the assumption of isentropic flow, the density relation and its derivative may become ill-defined for very high local velocities, which make the term in square brackets fall below zero. Because this corresponds to the physically impossible case of pressure below vacuum, the implementation simply clamps small values to a minimum pressure ratio of rho/rho_inf < 1E-5 . This limiter is often needed even for benign subsonic flow problems if there are regions of very high flow curvature, which may exist due to geometry (e.g., near wing-tip trailing edges) or insufficient mesh quality resulting in large dihedral angles between surface triangles. In these regions, the gradient of the potential can reach very large magnitudes, often only at a single mesh element."

@eloisabaez I would suggest keeping in mind and discussing the following points when the time comes for future PRs (after implementing the FindUpstreamElement):

- Using the approach presented in section 2.5 of Nishida instead of the current one
- Check if this approach is also necessary for other functions that have the same problem for high velocities (e.g. the local speed of sound and the local mach number)
- Move the computation of the density and its derivative to the utilities.

Reply:
Yes, I will keep this in mind.